### DISCUSSION TITLE
Form field

### DISCUSSION THREAD

**Discussion by andresz1** [adevinta/spark#759] [Decision making]
## Summary

After doing these two spikes:

- #682 
- #683 

I think the easiest way to have an API similar to the one that Chakra provides is building it from scratch. Most of the logic inside that Radix `Form` component is related to form validation which we do not want. I made a [Proof of Concept](https://github.com/adevinta/spark/pull/725) to illustrate how this could be done inspired by both Chakra and Radix.

## Proposed API

```jsx
<FormField name="password" isInvalid isRequired>
  <FormField.Label>Password</FormField.Label>

  <FormField.State>
    {({ id, name, description, isRequired, isInvalid }) => (
      <input
        type="password"
        id={id}
        name={name}
        aria-invalid={isInvalid}
        aria-describedby={description}
        aria-required={isRequired}
        required={isRequired}
        className="p-md border-neutral active:border-primary border-md rounded-sm"
      />
    )}
  </FormField.State>

  <FormField.HelperMessage>Never share your password</FormField.HelperMessage>

  <FormField.ErrorMessage>Password is required</FormField.ErrorMessage>
</FormField>
```

### Integration

Input components like `RadioGroup` should use `useFormFieldState` to integrate `FormField` out of the box. Since wrapping the component with `FormField` is **optional** the usage of `useFormFieldState` **won't throw any error** instead will return an empty object.

```jsx
<form>
  <FormControl name="condition">
    <FormLabel asChild>
      <p>Apparel condition</p>
    </FormLabel>

    <RadioGroup>
      <Radio value="1">New</Radio>
      <Radio value="2">Very good</Radio>
      <Radio value="3">Good</Radio>
      <Radio value="4">Satisfactory</Radio>
    </RadioGroup>

    <FormHelperMessage>The condition that best matches your product</FormHelperMessage>

    <FormErrorMessage>The condition is required</FormErrorMessage>
  </FormControl>
</form>
```

**Comment by andresz1** [comment]
I created a PR to implement the `FormControl` https://github.com/adevinta/spark/pull/758

**Comment by soykje** [comment]
@andresz1 Could it be possible to have more details about Spark form components (switch, radio, ...) implementing FormField? If I'm correct, it's simply about adding the `useFormFieldState()` hook where relevant, and adding some documentation about it?

**Reply by soykje** [reply]
Just opened a PR for `Switch` implementation: https://github.com/adevinta/spark/pull/782. If I understood correctly it should just be a matter of accessibility and control, thx to this hook. With some complementary tests obviously :smiley:

**Comment by Powerplex** [comment]
Component is done. Closing.
